sql/sql_table.h

```diff
@@ -285,4 +286,61 @@ extern mysql_mutex_t LOCK_gdl;
 
 bool check_engine(THD *, const char *, const char *, HA_CREATE_INFO *);
 
+/** Looks for the same key as needle in a haystack array. Keys should have the
+same name, key parts count, position in index, etc.
+@param[in]	haystack	array of KEYs
+@param[in]	haystack_size	array size
+@param[in]	needle	function looks for key equivalent to this one
+@return	true if equivalent index was found and false otherwise */
+bool has_equivalent_index(const KEY *haystack, const uint haystack_size,
+                          const KEY &needle);
+
+/** Checks whether some column was ALTERed and it's possible to do it INSTANT.
+@param[in]	ha_alter_info	info about ALTER TABLE
+@return	true if at least one field was ALTERed and every field can be ALTERed
+INSTANT */
+bool instant_alter_column_type_possible(
+    const Alter_inplace_info *ha_alter_info);
+
+/** Checks whether no fields is ALTERed or every ALTERed field can be altered
+INSTANT.
+@param[in]	ha_alter_info	alter info
+@return	true if no field types are changed or every field can be changed
+instantly */
+bool instant_alter_column_type_possible_or_irrelevant(
+    const Alter_inplace_info *ha_alter_info);
+
+/* Checks that at least one unique index is dropped and added and
+every unique key was dropped and added with the same key parts.
+@param[in]	ha_alter_info	alter info
+@return	true if check succeded and false otherwise */
+bool instant_drop_add_unique_index_possible(
+    const Alter_inplace_info *ha_alter_info);
+
+/* Checks that if unique indexes are dropped and added every
+every unique key was dropped and added with the same key parts.
+@param[in]	ha_alter_info	alter info
+@return	true if check succeded and false otherwise */
+bool instant_drop_add_unique_index_possible_or_irrelevant(
+    const Alter_inplace_info *ha_alter_info);
+
+/** Checks whether PRIMARY KEY was dropped and PRIMARY KEY over the same
+key parts was created.
+@param[in]	ha_alter_info	alter info
+@result	true if condition succedded and false otherwise */
+bool instant_drop_add_pk_possible(const Alter_inplace_info *ha_alter_info);
+
+/** Checks whether PRIMARY KEY was dropped and PRIMARY KEY over the same
+key parts was created.
+@param[in]	ha_alter_info	alter info
+@result	true if condition succedded and false otherwise */
+bool instant_drop_add_pk_possible_or_irrelevant(
+    const Alter_inplace_info *ha_alter_info);
+
+/** Checks whether it's possible to change charset of all fields instantly.
+@param[in]	ha_alter_info	alter_info
+@result	true if condition succedded and false otherwise */
+bool instant_alter_charset_possible_or_irrelevant(
+    const Alter_inplace_info *ha_alter_info);
+
```

Comment:
I think that these predicates would better be declared as member functions of `class Alter_inplace_info` in `handler.h`.